### ISSUE TITLE
Provide Secure Endpoints - Phase (implementation)

### ISSUE THREAD

**Issue by marikaj123** [kabanero-io/roadmap#150]
Kabanero should ensure that all networking between components can be over secure protocols.

**Comment by marikaj123** [comment]
Provide:  
 CLI service is called by CLI client

This is going to have multiple parts to get to where we need to be. Rather than have OpenShift trust the self-signed cert that Liberty uses, we're going to have Liberty use a certificate generated by OpenShift.
Configure OpenShift to generate a certificate into the landing page pod. See https://docs.openshift.com/container-platform/4.2/authentication/certificates/service-serving-certificate.html. For an example of how we did this with kabanero-eventing, see https://github.com/kabanero-io/kabanero-operator/blob/master/config/orchestrations/events/0.1/kabanero-events.yaml.
Convert the certificate to a keystore. See https://developers.redhat.com/blog/2017/11/22/dynamically-creating-java-keystores-openshift/. The certificate from step 1 should be mounted in the file system and we should use the keytool command to create a keystore from it. We will need to consider what container should build the keystore. We could use an init container as the example does, or do it right in the main container before we start Liberty. If we do it in the init container we must choose carefully what image we use, or build our own. The UBI base image (but not UBI-min) has openSSL, and the landing page image should have keytool I think but someone would need to verify that.
Update the liberty configuration for the landing page to use the keystore that was built in step 2.

**Comment by kaczyns** [comment]
WRT the previous comment - this work is already done.  The operator is generating a service serving certificate for the landing page, and an init container is re-packaging this as a pkcs12 file that the Liberty server is using.  The work was done before that comment was written, so, I'm wondering if there was a mistake.  Was some other component supposed to start using a service serving certificate?

**Comment by alohr51** [comment]
@kaczyns I talked with @jtmulvey  and for the landing/console app this work is done in https://github.com/kabanero-io/kabanero-operator/issues/178 (Marika added this issue to this epic for tracking purposes). The cert you generated from issue 178 is for the console app so others (like a browser) can trust this app, is that correct?

If the above assumption is correct, then I believe the CLI services app will have to do something similar because the console and the CLI client will both need to import a CLI Services cert so they can send a trusted/secure http request to it.

**Comment by marikaj123** [comment]
Create a new Issue  roadmap as a child of this Epic. 
Chunlong provided the description listed below, hoping this would assisst you in creating the appropriate child issues for this Epic: 
Here is a description you can use for any client that invokes appstack hub index

All client applications that invoke appstack hub index pages must be updated to support TLS and authentication requirement.
1. Update clients to support TLS.
Client must invoke appstack hub index over TLS, and must be able to accept server's certificate (you may just need add server certificate to client's trust store).
You can get server certificate by issuing the oc command below to obtain the root certificate for the ingress operator.
oc -n openshift-ingress-operator get secret router-ca -o=jsonpath='{.data.tls\.crt}'|base64 --decode | tee -a ingress.cert 
The certificate is displayed including -----BEGIN CERTIFICATE----- and the lines up to and through the -----END CERTIFICATE-----
You then add ingress.cert  to your client's trust store.
You can also use other tools, for example, open SSL,  to get server certificate.
2. Update clients to authenticate itself with the appstack hub index.
Client applications must request a service account token for kabanero namespace, and add the token as Bearer token. Here are steps used to request a service account token for authentication:
a. Create a service account token by issuing the following oc create sa kabanero-index-client-sa
b. Assign the service account the cluster view role oc -n kabanero adm policy add-role-to-user view -z kabanero-index-client-sa
c. Get the service account token oc serviceaccounts get-token kabanero-index-client-sa > satoken
d. add service account token to Authorization header as Bearer token. The header looks like:
	"Authorization: Bearer <service account token here>"
Here is a sample Curl request that supports TLS and service account token authentication
curl -v -X GET --cacert ./ingress.pem -H "Authorization: Bearer $sat" -H "Accept: application/json" https://kabanero-index-kabanero.apps.hostname.os.fyre.ibm.com/

**Comment by marikaj123** [comment]
Doc for Issue 70 may be needed.
